pinot-core/src/main/java/org/apache/pinot/core/indexsegment/immutable/ImmutableSegmentImpl.java

```diff
@@ -50,25 +51,29 @@
   private final SegmentMetadataImpl _segmentMetadata;
   private final Map<String, ColumnIndexContainer> _indexContainerMap;
   private final StarTreeIndexContainer _starTreeIndexContainer;
-  private final PartitionUpsertMetadataManager _partitionUpsertMetadataManager;
-  private final ValidDocIndexReader _validDocIndex;
+
+  // For upsert
+  private PartitionUpsertMetadataManager _partitionUpsertMetadataManager;
+  private ThreadSafeMutableRoaringBitmap _validDocIds;
+  private ValidDocIndexReader _validDocIndex;
 
   public ImmutableSegmentImpl(SegmentDirectory segmentDirectory, SegmentMetadataImpl segmentMetadata,
       Map<String, ColumnIndexContainer> columnIndexContainerMap,
-      @Nullable StarTreeIndexContainer starTreeIndexContainer,
-      @Nullable PartitionUpsertMetadataManager partitionUpsertMetadataManager) {
+      @Nullable StarTreeIndexContainer starTreeIndexContainer) {
     _segmentDirectory = segmentDirectory;
     _segmentMetadata = segmentMetadata;
     _indexContainerMap = columnIndexContainerMap;
     _starTreeIndexContainer = starTreeIndexContainer;
-    if (partitionUpsertMetadataManager != null) {
-      _partitionUpsertMetadataManager = partitionUpsertMetadataManager;
-      _validDocIndex =
-          new ValidDocIndexReaderImpl(partitionUpsertMetadataManager.createValidDocIds(getSegmentName()));
-    } else {
-      _partitionUpsertMetadataManager = null;
-      _validDocIndex = null;
-    }
+  }
+
+  /**
+   * Enables upsert for this segment.
+   */
+  public void enableUpsert(PartitionUpsertMetadataManager partitionUpsertMetadataManager,
+      ThreadSafeMutableRoaringBitmap validDocIds) {
+    _partitionUpsertMetadataManager = partitionUpsertMetadataManager;
+    _validDocIds = validDocIds;
+    _validDocIndex = new ValidDocIndexReaderImpl(validDocIds);
```

Comment:
Is it possible that this immutable segment is queried before the `enableUpsert ` is invoked?

If so, `_validDocIndex ` will be null and confuse the query plan

Reply:
No, the `enableUpsert` is called before adding the segment to the data manager.
Will add more javadoc stating that